### ISSUE TITLE
Multiple Contexts Behavior

### ISSUE THREAD

**Issue by francovitali** [jaystack/jaydata#45]
Provider: webSql
Version: 1.1.1

We use 2 contexts in our application to separate disctinct sets of data in two databases. 

We developed a configurable "toJSON" function, which retrieves all the hierarchy given an entity or set of entities. This code uses "getPublicMappedProperties()" to get the entity members and parse it recursively (see toJSON method implementation for a use example).

 [entity].memberDefinitions.getPublicMappedProperties()

The problem is that on some entities, we are getting members from entities from ANOTHER Context !!!!

In that particular case, the Entity and EntitySet were called the same on both Contexts. Also, the "ghost" properties were called the same. We changed the class names and now is working correclty.

We think this is Core Issue.

Does JayData support the use of more than one context? If not, please elaborate the reason, so we can contribute.

Thanks in advance!

**Comment by PeterAronZentai** [comment]
JayData _does_ support multiple concurrent contexts, in fact, it is the suggested way to isolate certain use cases.

We experienced issues on some mobile devices with two concurrent open connections to the same database. 

I tried to repro your case, and the following code did execute correctly without any kind of ghost property transfer (neither metadata nor sql table schema wise) with the latest build. Maybe it's too simple or the usage is too simple... Can you elaborate the environment? Is it a hybrid app? or a server side app? or a html5 app? :)

``` javascript
    $data.Class.define("$org.types.AccountType", $data.Entity, null, {
        Id : {
            type : "int",
            key : true,
            computed : true
        },
        Name : {
            type : "string",
            required : true
        }
    });

    $data.Class.define("$org2.types.AccountType", $data.Entity, null, {
        Id : {
            type : "int",
            key : true,
            computed : true
        },
        Age: {
            type : "number",
            required : true
        }
    });

    $data.EntityContext.extend("$org.MyContext", {
        AccountTypes: { type: $data.EntitySet, elementType: $org.types.AccountType }
    });

    $data.EntityContext.extend("$org2.MyContext", {
        AccountTypes: { type: $data.EntitySet, elementType: $org2.types.AccountType }
    });

    var c = new $org.MyContext({ name: 'sqLite', databaseName: 'mydb'});
    c.onReady(function(){
        var c2 = new $org2.MyContext({ name: 'sqLite', databaseName: 'mydb2'});
        c2.onReady(function(){
            c.AccountTypes.add({Name:"asdasd"});
            c.saveChanges( function() {
                alert("saved!");
            })

            c2.AccountTypes.add({Age:12});
            c2.saveChanges( function() {
                alert("saved!");
            })
        })

    })
```

**Comment by francovitali** [comment]
Thanks for the fast reply. We were actually doing things the wrong way.

Our Model Replicator was creating an individual file for each context, with an "inmendiate function" declaring the Entities, EntitySet and EntityContext.

Even if we waited for the contextReady events to fire, the 2 contexts were creating simultaneously.

Now we moved EntityContexts initialization code to another file and initialize the second EntityContext, inside the contextReady callback from the first.

This week we started testing in Android devices and the "simultaneous" initialization code throw all kind of errors. 

When i have time at work, i will try to reproduce the apperance of the "ghost" property again, but with the minor change, based on your code, everything is running smoothly right now.
